# Regressão II - parte 1

Regressão múltipla

#### Previsão de renda

Vamos trabalhar com a base 'previsao_de_renda.csv', que é a base do seu próximo projeto. Vamos usar os recursos que vimos até aqui nesta base.

|variavel|descrição|
|-|-|
|data_ref                | Data de referência de coleta das variáveis |
|index                   | Código de identificação do cliente|
|sexo                    | Sexo do cliente|
|posse_de_veiculo        | Indica se o cliente possui veículo|
|posse_de_imovel         | Indica se o cliente possui imóvel|
|qtd_filhos              | Quantidade de filhos do cliente|
|tipo_renda              | Tipo de renda do cliente|
|educacao                | Grau de instrução do cliente|
|estado_civil            | Estado civil do cliente|
|tipo_residencia         | Tipo de residência do cliente (própria, alugada etc)|
|idade                   | Idade do cliente|
|tempo_emprego           | Tempo no emprego atual|
|qt_pessoas_residencia   | Quantidade de pessoas que moram na residência|
|renda                   | Renda em reais|

In [301]:
# Carregando bibliotecas

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import patsy
import statsmodels.api as sm
import statsmodels.formula.api as smf

In [302]:
# Carregando dataset
df = pd.read_csv('previsao_de_renda.csv')
print(df.shape)
df.head(3)

(15000, 15)


Unnamed: 0    data_ref  id_cliente sexo  posse_de_veiculo  posse_de_imovel  \
0           0  2015-01-01       15056    F             False             True   
1           1  2015-01-01        9968    M              True             True   
2           2  2015-01-01        4312    F              True             True   

   qtd_filhos   tipo_renda           educacao estado_civil tipo_residencia  \
0           0   Empresário         Secundário     Solteiro            Casa   
1           0  Assalariado  Superior completo       Casado            Casa   
2           0   Empresário  Superior completo       Casado            Casa   

   idade  tempo_emprego  qt_pessoas_residencia    renda  
0     26       6.602740                    1.0  8060.34  
1     28       7.183562                    2.0  1852.15  
2     35       0.838356                    2.0  2253.89

In [303]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             15000 non-null  int64  
 1   data_ref               15000 non-null  object 
 2   id_cliente             15000 non-null  int64  
 3   sexo                   15000 non-null  object 
 4   posse_de_veiculo       15000 non-null  bool   
 5   posse_de_imovel        15000 non-null  bool   
 6   qtd_filhos             15000 non-null  int64  
 7   tipo_renda             15000 non-null  object 
 8   educacao               15000 non-null  object 
 9   estado_civil           15000 non-null  object 
 10  tipo_residencia        15000 non-null  object 
 11  idade                  15000 non-null  int64  
 12  tempo_emprego          12427 non-null  float64
 13  qt_pessoas_residencia  15000 non-null  float64
 14  renda                  15000 non-null  float64
dtypes:

1. Ajuste um modelo para prever log(renda) considerando todas as covariáveis disponíveis.
    - Utilizando os recursos do Patsy, coloque as variáveis qualitativas como *dummies*.
    - Mantenha sempre a categoria mais frequente como casela de referência
    - Avalie os parâmetros e veja se parecem fazer sentido prático.

2. Remova a variável menos significante e analise:
    - Observe os indicadores que vimos, e avalie se o modelo melhorou ou piorou na sua opinião.
    - Observe os parâmetros e veja se algum se alterou muito.

3. Siga removendo as variáveis menos significantes, sempre que o *p-value* for menor que 5%. Compare o modelo final com o inicial. Observe os indicadores e conclua se o modelo parece melhor. 
    

# 1. 

In [306]:
# Removendo colunas desnecessárias à análise
df = df.drop(['Unnamed: 0', 'data_ref', 'id_cliente'], axis=1)
df.shape

(15000, 12)

In [307]:
# Removendo dados faltantes
print(df.isnull().sum())

sexo                        0
posse_de_veiculo            0
posse_de_imovel             0
qtd_filhos                  0
tipo_renda                  0
educacao                    0
estado_civil                0
tipo_residencia             0
idade                       0
tempo_emprego            2573
qt_pessoas_residencia       0
renda                       0
dtype: int64


In [308]:
df = df.dropna()
print(df.shape)
df.isna().sum()

(12427, 12)


sexo                     0
posse_de_veiculo         0
posse_de_imovel          0
qtd_filhos               0
tipo_renda               0
educacao                 0
estado_civil             0
tipo_residencia          0
idade                    0
tempo_emprego            0
qt_pessoas_residencia    0
renda                    0
dtype: int64

In [309]:
# Variáveis qualitativas

df.select_dtypes('object').describe().transpose()

count unique          top   freq
sexo             12427      2            F   7901
tipo_renda       12427      5  Assalariado   7633
educacao         12427      5   Secundário   7045
estado_civil     12427      5       Casado   8897
tipo_residencia  12427      6         Casa  11071

#### <span style='color:blue'> Gerando o modelo

In [311]:
# Gerando matriz design e y

y, x = patsy.dmatrices('np.log(renda) ~ C(sexo) + posse_de_veiculo + posse_de_imovel + qtd_filhos + C(tipo_renda)' 
                       '+ C(educacao) + C(estado_civil) + C(tipo_residencia) + idade + tempo_emprego + qt_pessoas_residencia',
                       data = df)

reg = sm.OLS(y, x).fit()
reg.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.357
Model:                            OLS   Adj. R-squared:                  0.356
Method:                 Least Squares   F-statistic:                     287.5
Date:                Tue, 29 Oct 2024   Prob (F-statistic):               0.00
Time:                        17:36:59   Log-Likelihood:                -13568.
No. Observations:               12427   AIC:                         2.719e+04
Df Residuals:                   12402   BIC:                         2.737e+04
Df Model:                          24                                         
Covariance Type:            nonrobust                                         
=======================================================================================================
                                          coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------
Intercept                               6.5847      0.235     28.006      0.000       6.124       7.046
C(sexo)[T.M]                            0.7874      0.015     53.723      0.000       0.759       0.816
posse_de_veiculo[T.True]                0.0441      0.014      3.119      0.002       0.016       0.072
posse_de_imovel[T.True]                 0.0829      0.014      5.926      0.000       0.055       0.110
C(tipo_renda)[T.Bolsista]               0.2209      0.241      0.916      0.360      -0.252       0.694
C(tipo_renda)[T.Empresário]             0.1551      0.015     10.387      0.000       0.126       0.184
C(tipo_renda)[T.Pensionista]           -0.3087      0.241     -1.280      0.201      -0.782       0.164
C(tipo_renda)[T.Servidor público]       0.0576      0.022      2.591      0.010       0.014       0.101
C(educacao)[T.Pós graduação]            0.1071      0.159      0.673      0.501      -0.205       0.419
C(educacao)[T.Secundário]              -0.0141      0.072     -0.196      0.844      -0.155       0.127
C(educacao)[T.Superior completo]        0.0938      0.072      1.298      0.194      -0.048       0.235
C(educacao)[T.Superior incompleto]     -0.0432      0.078     -0.555      0.579      -0.196       0.109
C(estado_civil)[T.Separado]             0.3265      0.111      2.935      0.003       0.108       0.545
C(estado_civil)[T.Solteiro]             0.2672      0.109      2.453      0.014       0.054       0.481
C(estado_civil)[T.União]               -0.0343      0.025     -1.361      0.174      -0.084       0.015
C(estado_civil)[T.Viúvo]                0.3689      0.116      3.191      0.001       0.142       0.596
C(tipo_residencia)[T.Casa]             -0.0442      0.054     -0.815      0.415      -0.150       0.062
C(tipo_residencia)[T.Com os pais]      -0.0258      0.060     -0.427      0.670      -0.144       0.093
C(tipo_residencia)[T.Comunitário]      -0.1196      0.105     -1.135      0.256      -0.326       0.087
C(tipo_residencia)[T.Estúdio]           0.0665      0.099      0.670      0.503      -0.128       0.261
C(tipo_residencia)[T.Governamental]    -0.0571      0.066     -0.865      0.387      -0.186       0.072
qtd_filhos                             -0.2658      0.108     -2.463      0.014      -0.477      -0.054
idade                                   0.0053      0.001      6.624      0.000       0.004       0.007
tempo_emprego                           0.0617      0.001     59.398      0.000       0.060       0.064
qt_pessoas_residencia                   0.2926      0.108      2.716      0.007       0.081       0.504
==============================================================================
Omnibus:                        0.858   Durbin-Watson:                   2.023
Prob(Omnibus)

# 2.

Para o modelo em questão, a variável menos significativa é "tipo_residencia". Vamos removê-lá.

In [313]:
# Removendo a variável menos significante "tipo_residencia"

reg = smf.ols('np.log(renda) ~ C(sexo) + posse_de_veiculo + posse_de_imovel + qtd_filhos + C(tipo_renda)' 
                       '+ C(educacao) + C(estado_civil) + idade + tempo_emprego + qt_pessoas_residencia',
                       data = df).fit()
reg.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.357
Model:                            OLS   Adj. R-squared:                  0.356
Method:                 Least Squares   F-statistic:                     363.0
Date:                Tue, 29 Oct 2024   Prob (F-statistic):               0.00
Time:                        17:36:59   Log-Likelihood:                -13569.
No. Observations:               12427   AIC:                         2.718e+04
Df Residuals:                   12407   BIC:                         2.733e+04
Df Model:                          19                                         
Covariance Type:            nonrobust                                         
======================================================================================================
                                         coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------
Intercept                              6.5411      0.229     28.532      0.000       6.092       6.991
C(sexo)[T.M]                           0.7893      0.015     53.999      0.000       0.761       0.818
posse_de_veiculo[T.True]               0.0442      0.014      3.125      0.002       0.016       0.072
posse_de_imovel[T.True]                0.0819      0.014      5.966      0.000       0.055       0.109
C(tipo_renda)[T.Bolsista]              0.2219      0.241      0.920      0.357      -0.251       0.695
C(tipo_renda)[T.Empresário]            0.1556      0.015     10.433      0.000       0.126       0.185
C(tipo_renda)[T.Pensionista]          -0.3116      0.241     -1.292      0.196      -0.784       0.161
C(tipo_renda)[T.Servidor público]      0.0589      0.022      2.653      0.008       0.015       0.102
C(educacao)[T.Pós graduação]           0.1104      0.159      0.695      0.487      -0.201       0.422
C(educacao)[T.Secundário]             -0.0097      0.072     -0.135      0.893      -0.150       0.131
C(educacao)[T.Superior completo]       0.0980      0.072      1.359      0.174      -0.043       0.239
C(educacao)[T.Superior incompleto]    -0.0384      0.078     -0.495      0.621      -0.191       0.114
C(estado_civil)[T.Separado]            0.3272      0.111      2.942      0.003       0.109       0.545
C(estado_civil)[T.Solteiro]            0.2683      0.109      2.464      0.014       0.055       0.482
C(estado_civil)[T.União]              -0.0329      0.025     -1.310      0.190      -0.082       0.016
C(estado_civil)[T.Viúvo]               0.3678      0.116      3.182      0.001       0.141       0.594
qtd_filhos                            -0.2655      0.108     -2.460      0.014      -0.477      -0.054
idade                                  0.0052      0.001      6.579      0.000       0.004       0.007
tempo_emprego                          0.0617      0.001     59.485      0.000       0.060       0.064
qt_pessoas_residencia                  0.2923      0.108      2.713      0.007       0.081       0.504
==============================================================================
Omnibus:                        0.825   Durbin-Watson:                   2.023
Prob(Omnibus):                  0.662   Jarque-Bera (JB):                0.809
Skew:                           0.019   Prob(JB):                        0.667
Kurtosis:                       3.009   Cond. No.                     2.16e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.16e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

<span style='color:purple'>**Note que ao remover a variável "tipo_residencia" os valores de $R^{2}$ e `R-quadrado ajustado` nem mesmo foram alterados. Já o valor de `AIC` diminuiu um pouco ($[2.719e+04] \rightarrow [2.718e+04]$), o que indica uma melhora do modelo.**

# 3.

Siga removendo as variáveis menos significantes, sempre que o *p-value* for menor que 5%. Compare o modelo final com o inicial. Observe os indicadores e conclua se o modelo parece melhor. 


- De acordo com a análise, podemos remover as variáveis: `educacao`, `estado_civil` e `qt_pessoas_residencia`. A variável `qt_pessoas_residencia` apesar de possuir um bom p-value possui um intervalo de confiança muito grande, dessa forma sua remoção não afetará o modelo.
- A variável `tipo_renda` possui alguns valores de p-value e intervalos de confiança questionáveis (`C(tipo_renda)[T.Bolsista]` e `C(tipo_renda)[T.Pensionista]`), mas também possui "subvariáveis" com bons valores de p-value e intervalos de confiança (`C(tipo_renda)[T.Empresário]` e `C(tipo_renda)[T.Servidor público]`), logo será mantida no modelo. 

In [316]:
reg = smf.ols('np.log(renda) ~ C(sexo) + posse_de_veiculo + posse_de_imovel + qtd_filhos + C(tipo_renda)' 
                       '+ idade + tempo_emprego',
                       data = df).fit()

reg.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.353
Model:                            OLS   Adj. R-squared:                  0.353
Method:                 Least Squares   F-statistic:                     677.5
Date:                Tue, 29 Oct 2024   Prob (F-statistic):               0.00
Time:                        17:36:59   Log-Likelihood:                -13610.
No. Observations:               12427   AIC:                         2.724e+04
Df Residuals:                   12416   BIC:                         2.732e+04
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
=====================================================================================================
                                        coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------------
Intercept                             7.1456      0.034    207.748      0.000       7.078       7.213
C(sexo)[T.M]                          0.7790      0.015     53.600      0.000       0.751       0.807
posse_de_veiculo[T.True]              0.0537      0.014      3.833      0.000       0.026       0.081
posse_de_imovel[T.True]               0.0847      0.014      6.166      0.000       0.058       0.112
C(tipo_renda)[T.Bolsista]             0.2955      0.242      1.223      0.221      -0.178       0.769
C(tipo_renda)[T.Empresário]           0.1632      0.015     10.977      0.000       0.134       0.192
C(tipo_renda)[T.Pensionista]         -0.2676      0.241     -1.108      0.268      -0.741       0.206
C(tipo_renda)[T.Servidor público]     0.0769      0.022      3.475      0.001       0.034       0.120
qtd_filhos                            0.0288      0.009      3.349      0.001       0.012       0.046
idade                                 0.0053      0.001      6.864      0.000       0.004       0.007
tempo_emprego                         0.0616      0.001     59.325      0.000       0.060       0.064
==============================================================================
Omnibus:                        1.091   Durbin-Watson:                   2.024
Prob(Omnibus):                  0.579   Jarque-Bera (JB):                1.060
Skew:                           0.019   Prob(JB):                        0.589
Kurtosis:                       3.025   Cond. No.                     1.58e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.58e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

#### Insights

- Comparando os modelos podemos concluir que após a remoção das variáveis menos significativas, os valores de $R^{2}$ e *Adj. R-quadrado* diminuíram indicando que o modelo piorou em qualidade.
- O valor associado ao parâmetro `AIC` aumentou e isso indica uma piora na qualidade do modelo.
- Um ponto a se destacar é que o modelo final, apesar de possuir indicadores um pouco piores, é capaz de explicar bem a variabilidade dos dados, se comparado ao primeiro modelo (`R-quadrado` partiu de 0.357, no primeiro modelo, para 0.353, no último). Logo, o modelo final tem bons valores e não está sobrecarregado de variáveis pouco explicativas, evitando assim um possível overfitting.